tutorial/lesson_19_wrapper_funcs.cpp

```diff
@@ -0,0 +1,405 @@
+        // shared GPU memory.
+
+        // Consider a pipeline that transposes some compute_root'd Func:
+
+        Func f("f_transpose"), g("g_transpose");
+        f(x, y) = sin(((x + y) * sqrt(y)) / 10);
+        f.compute_root();
+
+        g(x, y) = f(y, x);
+
+        // The execution strategy we want is to load an 4x4 tile of f
+        // into registers, transpose it in-register, and then write it
+        // out as an 4x4 tile of g. We will use Func::in to express this:
+
+        Func f_tile = f.in(g);
+
+        // We now have a three stage pipeline:
+        // f -> f_transpose -> g
+
+        // f_tile will be responsible for loading from f into
+        // registers, f_tile_transpose will load vectors of f, and
+        // store them transposed. g will then write this data back to
+        // main memory.
+        g.tile(x, y, xo, yo, xi, yi, 4, 4)
+            .vectorize(xi)
+            .unroll(yi);
+
+        // We will compute f_transpose at tiles of g, and use
+        // Func::reorder_storage to state that f_transpose should be
+        // stored column-major, so that the loads to it done by g can
+        // be dense vector loads.
+        f_tile.compute_at(g, xo)
+            .reorder_storage(y, x)
+            .vectorize(x)
+            .unroll(y);
+
+        // We take care to make sure f_transpose is only ever accessed
+        // at constant indicies. The full unrolling/vectorization of
+        // all loops that exist inside its compute_at level has this
+        // effect. Allocations that are only ever accessed at constant
+        // indices can be promoted into registers.
+
+        g.realize(16, 16);
+        // See figures/lesson_19_transpose.mp4 for a visualization
+    }
+
+    {
+        // ImageParam::in behaves the same way as Func::in, and you
+        // can use it to stage loads in similar ways. Instead of
+        // transposing again, we'll use ImageParam::in to stage tiles
+        // of an input image into GPU shared memory, effectively using
+        // shared/local memory as an explicitly-managed cache.
+
+        ImageParam img(Int(32), 2);
+
+        // We will compute a small blur of the input.
+        Func blur("blur");
+        blur(x, y) = (img(x - 1, y - 1) + img(x, y - 1) + img(x + 1, y - 1) +
+                      img(x - 1, y    ) + img(x, y    ) + img(x + 1, y    ) +
+                      img(x - 1, y + 1) + img(x, y + 1) + img(x + 1, y + 1));
+
+        blur.compute_root().gpu_tile(x, y, xo, yo, xi, yi, 8, 8);
+
+        // The wrapper Func created by ImageParam::in has pure vars
+        // named _0, _1, etc. Schedule it per tile of "blur", and map
+        // _0 and _1 to gpu threads.
+        img.in(blur).compute_at(blur, xo).gpu_threads(_0, _1);
+
+        // Without Func::in, an 8x8 tile of f would do 8x8x9 loads to
+        // global memory. With Func::in, the wrapper does 10x10 loads
+        // to global memory up front, and then f does 8x8x9 loads to
```

Comment:
Perhaps use `*` instead of `x` here for clarity. I feel silly asking for this in prose when \times does seem appropriate, but I was confused by this at first since it sounded like it was loading an 8x8x9 block or something, not doing 8*8*9 = 576 loads.